server/accounts.go

```diff
@@ -3281,3 +3571,185 @@ func (dr *CacheDirAccResolver) Start(s *Server) error {
 	s.Noticef("Managing some jwt in exclusive directory %s", dr.directory)
 	return nil
 }
+
+func (dr *CacheDirAccResolver) Reload() error {
+	return dr.DirAccResolver.Reload()
+}
+
+// Transforms for arbitrarily mapping subjects from one to another for maps, tees and filters.
+// These can also be used for proper mapping on wildcard exports/imports.
+// These will be grouped and caching and locking are assumed to be in the upper layers.
+type transform struct {
+	src, dest string
+	dtoks     []string
+	stoks     []string
+	dtpi      []int8
+}
+
+// Helper to pull raw place holder index. Returns -1 if not a place holder.
+func placeHolderIndex(token string) int {
+	if len(token) > 1 && token[0] == '$' {
+		var tp int
+		if n, err := fmt.Sscanf(token, "$%d", &tp); err == nil && n == 1 {
+			return tp
+		}
+	}
+	return -1
+}
+
+// newTransform will create a new transform checking the src and dest subjects for accuracy.
+func newTransform(src, dest string) (*transform, error) {
+	// Both entries need to be valid subjects.
+	sv, stokens, npwcs, hasFwc := subjectInfo(src)
+	dv, dtokens, dnpwcs, dHasFwc := subjectInfo(dest)
+
+	// Make sure both are valid, match fwc if present and there are no pwcs in the dest subject.
+	if !sv || !dv || dnpwcs > 0 || hasFwc != dHasFwc {
+		return nil, ErrBadSubject
+	}
+
+	var dtpi []int8
+
+	// If the src has partial wildcards then the dest needs to have the token place markers.
+	if npwcs > 0 || hasFwc {
+		// We need to count to make sure that the dest has token holders for the pwcs.
+		sti := make(map[int]int)
+		for i, token := range stokens {
+			if len(token) == 1 && token[0] == pwc {
+				sti[len(sti)+1] = i
+			}
+		}
+
+		nphs := 0
+		for _, token := range dtokens {
+			tp := placeHolderIndex(token)
+			if tp >= 0 {
+				if tp > npwcs {
+					return nil, ErrBadSubject
+				}
+				nphs++
+				// Now build up our runtime mapping from dest to source tokens.
+				dtpi = append(dtpi, int8(sti[tp]))
+			} else {
+				dtpi = append(dtpi, -1)
+			}
+		}
+
+		if nphs != npwcs {
+			return nil, ErrBadSubject
+		}
+	}
+
+	return &transform{src: src, dest: dest, dtoks: dtokens, stoks: stokens, dtpi: dtpi}, nil
+}
+
+// match will take a literal published subject that is associated with a client and will match and transform
+// the subject if possible.
+// TODO(dlc) - We could add in client here to allow for things like foo -> foo.$ACCOUNT
```

Comment:
not sure if $ACCOUNT can be trusted (not faked) especially when used cross account

Reply:
Server fills it in, so can't be faked.. and you would own the mapping construct for the export.

Reply:
Care needs to be taken that if foo -> foo.$ACCOUNT exists, the user can't send foo.<not my account> themselves